### ISSUE TITLE
Waterfill explosive erases spidertron from existence

### ISSUE THREAD

**Issue by Nebisan** [Deadlock989/IndustrialRevolution#288]
**Describe the issue**
When a waterfill explosive detonates on a tile which is occupied by a spidertron's foot, the spidertron is instantly erased.
Footage: https://youtu.be/FqlI0wh4aRk
This happens regardless of if the spidertron is occupied or unoccupied. 

**Factorio version, IR version**
Factorio 1.1.87
IR 3.1.5

**Other mods installed**
I have a lot of mods installed, but all of them are small QOL/GUI/retexture mods. These are the ones I suspect may have some interaction with this bug.
- Alien biomes
- Car/tank/spidertron locator button
- Clean tiles
- Spidertron enhancements

I can provide a full list if necessary.

**Comment by Deadlock989** [comment]
That's interesting.

I can't reproduce this in IR3 standalone - waterfill explosives don't affect the spider when I set up a very similar situation to the one shown in your video. 

I also tried adding Spidertron Enhancements and there was no change, so you can rule that out. You can also rule out Clean Tiles (if it is [this mod](https://mods.factorio.com/mod/CleanTiles)) since that mod has no runtime scripting (you probably don't need it anyway, IR3 also enforces 100% decorative removal for vanilla placeable tiles).

That leaves Alien Biomes or one of your other mods. If you discover anything that might shed light on which, let me know, but my gut feeling is that this is not something I can "fix".

**Comment by Nebisan** [comment]
I've figured it out - it's alien biomes. Specifically, if I disable "include inland shallows" in the mod settings, then the spidertron bug occurs. It doesn't seem to matter whether the option is enabled during world generation, only if it's enabled at the moment of the waterfill going off.

**Comment by Deadlock989** [comment]
Thanks for the update. Some context: IR3 uses the vanilla tile `water-mud` for the waterfill explosive - this tile isn't used in map generation but is set up by the base mod to be traversable by characters, vehicles and enemy units (maybe for scenarios or tutorials, I don't know). The tile's `collision_mask` property is set up such that the tile is considerably closer to behaving like a ground tile than a water tile, i.e. the aforementioned entities do not collide with it.

I peeked into AB and it's not fully clear to me what is happening or what the intention was, but it seems (in lines 577-608 in `prototypes/biomes/biomes.lua`) that if the `alien-biomes-include-inland-shallows` setting is left on the default, then the `water-mud` tile ends up with a different and unusual `collision_mask` compared to vanilla's, but one that still does not collide with vehicles, which is fine for IR3's purposes and there is no issue. If however the setting is disabled, then some other code elsewhere in AB results in the tile getting the following `collision_mask` instead:

``{
  "water-tile",
  "item-layer",
  "resource-layer",
  "object-layer",
  "item-layer"
}``

This is actually the same collection of masks as the vanilla default for `water-mud` that IR3 expects, except that `item-layer` is duplicated. I don't know if that's really the cause of this issue due to some weirdness in the game engine when collision layers are accidentally specified twice, or if this is just a red herring and there is something else going on with AB and spidertrons, but at any rate I can confirm that spidertrons are destroyed when a tile with this particular collision mask is spawned underneath them - if I recall correctly that is the normal behaviour for an entity when you create a colliding tile under it via scripts or the built-in editor.

In theory I could try to update the tile's collision mask at the last possible moment in order to try and make sure that it has the desired vanilla default. But AB isn't directly supported by IR3 - this is something AB's author(s) would have to address if you really want to use the non-default setting.

**Comment by tburrows13** [comment]
Hello, I can shed some more light on this one.
- When AB is installed with default settings, `water-mud` doesn't have `object-layer`. When you disable "include inland shallows", then `water-mud` gets given `object-layer`. No idea why.
- **AB adds `object-layer` to spider legs**. Also no idea why, I think it is a legacy thing that I've noticed caused problems with other mods too. Now that I know it causes problems with Alien Biomes itself, I'm going to try and get it removed.
- As you said, entities being deleted when a colliding tile is created under them is standard engine behaviour. There are 2 properties that exist, which can prevent this behaviour when placing tiles from an item (e.g. waterfill items): [tile.check_collision_with_entities](https://lua-api.factorio.com/latest/prototypes/TilePrototype.html#check_collision_with_entities) and [entity. protected_from_tile_building ](https://lua-api.factorio.com/latest/prototypes/EntityPrototype.html#protected_from_tile_building). These don't apply when script creating tiles so your waterfill explosives aren't protected by them. I guess you _could_ try and detect colliding entities and not place waterfill there, but that is probably outside the scope of what is reasonable.

In summary, I don't think you really need to bother with 'fixing' this :)

(By the way, I think the waterfill explosive mechanic is awesome, so much cooler than just placing water tiles!)

**Comment by Deadlock989** [comment]
Ah. Yeah, that would do it.
